server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.routing;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.opensearch.OpenSearchException;
+import org.opensearch.action.search.SearchShardIterator;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.WeightedRoutingMetadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.rest.RestStatus;
+import org.opensearch.search.SearchShardTarget;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * This class contains logic to find next shard to retry search request in case of failure from other shard copy.
+ * This decides if retryable shard search requests can be tried on shard copies present in data
+ * nodes whose attribute value weight for weighted shard routing is set to zero.
+ */
+
+public class FailAwareWeightedRouting {
+
+    private static final Logger logger = LogManager.getLogger(FailAwareWeightedRouting.class);
+
+    /* exception causing failure for previous shard copy */
+    private Exception exception;
+
+    private ClusterState clusterState;
+
+    private final static List<RestStatus> internalErrorRestStatusList = List.of(
+        RestStatus.INTERNAL_SERVER_ERROR,
+        RestStatus.NOT_IMPLEMENTED,
+        RestStatus.BAD_GATEWAY,
+        RestStatus.SERVICE_UNAVAILABLE,
+        RestStatus.GATEWAY_TIMEOUT,
+        RestStatus.HTTP_VERSION_NOT_SUPPORTED,
+        RestStatus.INSUFFICIENT_STORAGE
+    );
+
+    public FailAwareWeightedRouting(Exception e, ClusterState clusterState) {
+        this.exception = e;
+        this.clusterState = clusterState;
+    }
```

Comment:
Please use ClusterService in the constructor so that it could be used as a singleton instance and move exception to the method `findNext` and log the same exception

Reply:
I think we can't make `FailAwareWeightedRouting` singleton since even cluster state can change between requests

Reply:
Yeah bad edit, I initially meant ClusterService for singleton but forgot to remove.

Reply:
Got it. I have made `FailAwareWeightedRouting` singleton but we can't pass `ClusterService` in the constructor since `ClusterService` instance is not available in `AbstractSearchAsyncAction`.